ServerCore/Pages/Submissions/Index.cshtml.cs

```diff
@@ -95,12 +102,14 @@ public async Task<IActionResult> OnPostAsync(int puzzleId, string submissionText
             }
 
             // Soft enforcement of duplicates to give a friendly message in most cases
-            DuplicateSubmission = (from sub in Submissions
+            Submission duplicatedSubmission = (from sub in Submissions
                                    where sub.SubmissionText == ServerCore.DataModel.Response.FormatSubmission(submissionText)
-                                   select sub).Any();
+                                   select sub).FirstOrDefault();
 
-            if (DuplicateSubmission)
+            if (duplicatedSubmission != null)
             {
+                string response = duplicatedSubmission.Response != null ? duplicatedSubmission.Response.ResponseText : IndexModel.IncorrectResponseText;
```

Comment:
This also doesn't handle freeform correctly; you should probably just have a message like "you already submitted this" if the puzzle is freeform.

Beyond that, this is a shorter way to write the non-freeform parts:

string response = duplicatedSubmission.Response?.ResponseText ?? IndexModel.IncorrectResponseText;

That will use IndexModel.IncorrectResponseText if either duplicatedSubmission.Response is null or duplicatedSubmission.Response.ResponseText is null.

Reply:
Fixed